Scatter reorganization splitting the logic into multiple classes/objects and using the mvvm pattern

airenzp opened the pull request:
## Description

The scatter plot now creates a ScatterView, scatterModel, ScatterViewModel and a ScatterInteractivity to handle the logic related to each area, following the mvvm pattern. The ScatterViewModel is inherited by ScatterViewModel2DLarge and ScatterViewModel3D that overwrite renderSeries, to render large plots and 3D plots. In order to further modularize the code logic the ScatterViewModel  also creates the ScatterLasso, ScatterTooltip, ScatterZoom and the ScatterLegend, that has its own ScatterLegendInteractivity. All the files are using typescript and the current tests are passing. 
I am planning to add further tests on the scatter, as planned, but I wanted to create this PR as there are so many files involved and the scatter is recreated from scratch, to give some time to test and gather feedback from you.


## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

xzhou82 commented on shared/types/src/dataset.ts:
```diff
@@ -1027,6 +1027,7 @@ this is limited to only one term and doesn't allow switching between multiple te
 		/** order of categories */
 		order: string[]
 	}
+	isLarge?: boolean
```
please add comment

airenzp replied:
This property was removed after noticing I needed to recreate the ScatterViewModel on main to, for example, render a 3d plot when a Z term is added in continuous mode. Now that the renderer is created after getting the data from the backend I can check the amount of particles to see if it is a large plot or not.

creilly8 commented on client/plots/scatter/Scatter.ts:
```diff
@@ -15,7 +15,20 @@ import { controlsInit } from '../controls'
+	charts: any
+	opts: any
+	state: any
+	type: string
```
It maybe easier to declare the `type` as const here. Any immutable variable can be declared as `readonly myVar = 'value'` and the type is inferred. It makes it easier to debug later.

creilly8 commented on client/plots/scatter/Scatter.ts:
```diff
@@ -15,7 +15,20 @@ import { controlsInit } from '../controls'
+	settings: any
+	charts: any
+	opts: any
+	state: any
```
As a reference to make your life easier, there are mass specific, base types already available in `client/mass/types/mass.ts`. When you make types for these later, you can use those base types to build on.

creilly8 commented on client/plots/scatter/viewmodel/scatterLasso.ts:
```diff
@@ -163,8 +170,8 @@ export class ScatterLasso {
-		const rows = []
-		const columns = []
+		const rows: any[] = []
+		const columns: any = []
```
For these two you can do 
```
import type { TableCell, TableColumn } from '#dom'
......
     const rows: TableCell[][] = []
     const columns: TableColumn[] = []
.....
```

creilly8 commented on client/plots/scatter/model/scatterModel.ts:
```diff
@@ -0,0 +1,369 @@
+import { shapesArray } from '#dom'
+import { rgb } from 'd3-color'
+import { scaleLinear as d3Linear } from 'd3-scale'
```
I think `scaleTime` is missing in this import statement

creilly8 commented:
@airen. I tested the examples on http://localhost:3000/url.html. Here's some of the issues I noticed: 
1. When the height and width of the plot is changed, the main div expands to size but not the plot. To reproduce, change the height and width in [this PNET example](http://localhost:3000/?mass={%22dslabel%22:%22PNET%22,%22genome%22:%22hg19%22,%22nav%22:{%22activeTab%22:1},%22plots%22:[{%22chartType%22:%22sampleScatter%22,%22name%22:%22Methylome%20TSNE%22}]}) with the axes shown. The axis show the main div responds but not the plot on dispatch. Adding a divide by term is the only way I noticed the plot corrects to the right size. 
2. I tried to change the max color in the color scale with [this PNET example](http://localhost:3000/?mass={%22dslabel%22:%22PNET%22,%22genome%22:%22hg19%22,%22nav%22:{%22activeTab%22:1},%22plots%22:[{%22chartType%22:%22sampleScatter%22,%22name%22:%22Methylome%20TSNE%22,%22colorTW%22:{%22id%22:%22Age%22,%22q%22:{%22mode%22:%22continuous%22}},%22shapeTW%22:{%22id%22:%22chr1p%22}}]}). The error I see in the console is ```Uncaught (in promise) TypeError: this.changeGradientColor is not a function```. No issues with changing the min/max from that menu. 
3. I see the issue with the legend cutoff we've been chatting back and forth in [this PNET example](http://localhost:3000/?mass={%22dslabel%22:%22PNET%22,%22genome%22:%22hg19%22,%22plots%22:[{%22chartType%22:%22sampleScatter%22,%22name%22:%22Methylome%20TSNE%22,%22colorTW%22:{%22id%22:%22TSNE%20Category%22},%22shapeTW%22:{%22term%22:{%22gene%22:%22KRAS%22,%22type%22:%22geneVariant%22}}}]}). It looks like this for me. 
<img width="159" alt="Screenshot 2025-04-30 at 11 27 01 AM" src="https://github.com/user-attachments/assets/23bac1ea-05b7-425d-8628-a0a7a827ecfc" />  Including `"nav":{"activeTab":-1}` in the URL fixed it for me - which is echoes the explanation you mentioned in https://github.com/stjude/proteinpaint/issues/3218. We can talk about this more if you like. 

Besides this, all of the advanced functions appear intact after this reorg. I did not see any issues with adding terms, showing regression lines, contour maps, etc. from the examples I tested.

airenzp commented:
I think I have fixed the issues 1 and 2 @creilly8 in my last commit. Please check. Regarding the 3rd issue, the idea is not to create links with predefined scatter plots that by default open in the about tab, otherwise this clipping happens as we discussed. These urls should land on the plots tab anyways. I updated the scatter links that I found, basically adding the activeTab:1 in the params. Where is the url that you shared used? If we add the activeTab:1 there the legend gets fixed, see [url](http://localhost:3000/?mass=%7B%22dslabel%22:%22PNET%22,%22genome%22:%22hg19%22,%22nav%22:{%22activeTab%22:1},%22plots%22:%5B%7B%22chartType%22:%22sampleScatter%22,%22name%22:%22Methylome%20TSNE%22,%22colorTW%22:%7B%22id%22:%22TSNE%20Category%22%7D,%22shapeTW%22:%7B%22term%22:%7B%22gene%22:%22KRAS%22,%22type%22:%22geneVariant%22%7D%7D%7D%5D%7D).

creilly8 commented:
> I think I have fixed the issues 1 and 2 @creilly8 in my last commit. Please check. Regarding the 3rd issue, the idea is not to create links with predefined scatter plots that by default open in the about tab, otherwise this clipping happens as we discussed. These urls should land on the plots tab anyways. I updated the scatter links that I found, basically adding the activeTab:1 in the params. Where is the url that you shared used? If we add the activeTab:1 there the legend gets fixed, see [url](http://localhost:3000/?mass=%7B%22dslabel%22:%22PNET%22,%22genome%22:%22hg19%22,%22nav%22:%7B%22activeTab%22:1%7D,%22plots%22:%5B%7B%22chartType%22:%22sampleScatter%22,%22name%22:%22Methylome%20TSNE%22,%22colorTW%22:%7B%22id%22:%22TSNE%20Category%22%7D,%22shapeTW%22:%7B%22term%22:%7B%22gene%22:%22KRAS%22,%22type%22:%22geneVariant%22%7D%7D%7D%5D%7D).

I see the issues noted above in #1 and #2 fixed. Thanks! As for #3, the url is a scatter plot example on url.html. I'm not aware if it's used anywhere, such as an active item or default plot in `massnative`, anywhere. .

airenzp commented:
I see, I am adding a fix for that link in this [PR](https://github.com/stjude/sjpp/pull/809). Thanks Colleen!

xzhou82 commented:
minor issue. on switching to wilcoxon, `Loading..` doesn't show

xzhou82 reviewed (approved):
thanks a lot
